index/segment/mem/searcher.go

```diff
@@ -74,5 +111,62 @@ func (s *sequentialSearcher) Query(query segment.Query) ([]doc.Document, error)
 	}
 
 	// TODO: once we support multiple segments, we'll have to merge results
-	return s.queryable.Fetch(candidateDocIds)
+	return s.queryable.Fetch(candidateDocIds, matchPredicates(predicates).Fn())
+}
+
+type matchPredicates []matchPredicate
+
+func (m matchPredicates) Fn() matchPredicate {
+	return func(d doc.Document) bool {
+		for _, fn := range m {
+			if !fn(d) {
+				return false
+			}
+		}
+		return true
+	}
+}
+
+// orderFiltersByNonNegated orders filters which are not negated first in the list.
+type orderFiltersByNonNegated []segment.Filter
+
+func (of orderFiltersByNonNegated) Len() int           { return len(of) }
+func (of orderFiltersByNonNegated) Swap(i, j int)      { of[i], of[j] = of[j], of[i] }
+func (of orderFiltersByNonNegated) Less(i, j int) bool { return !of[i].Negate && of[j].Negate }
+
+// validate any assumptions we have about queries.
+func validateQuery(q segment.Query) error {
+	// assuming we only support AndConjuctions for now.
+	if q.Conjunction != segment.AndConjunction {
```

Comment:
Not an issue now, but is it worth calling out that we probably won't support union with negation filters? i.e. queries such as `key1=value1 OR key2!=value2` will be illegal since it's a superset of `key2!=value2` which is illegal